runtime/near-vm-runner/src/cache.rs

```diff
@@ -28,41 +53,72 @@ pub(crate) fn compile_module_cached(
-        return compile_module(hash.clone(), code, config);
+        return compile_module(code, config);
     }
+    let key = get_key(code_hash, code, config);
     let cache = cache.unwrap();
-    if let Some(serialized) = cache
-        .get(&hash)
-        .map_err(|_e| VMError::CacheError(format!("Cannot read from cache {:?}", &code_hash)))?
-    {
-        let artifact = Artifact::deserialize(&serialized).map_err(|_e| {
-            VMError::CacheError(format!("Cannot deserialize from cache {:?}", &code_hash))
-        })?;
-        unsafe {
-            load_cache_with(artifact, compiler_for_backend(Backend::Singlepass).unwrap().as_ref())
-                .map_err(|_e| {
-                    VMError::CacheError(format!("Cannot read from cache {:?}", &code_hash))
-                })
-        }
-    } else {
-        // TODO: cache compilation errors as well. Currently it's possible to deploy invalid code,
-        // and compile on every call
-        let compiled: wasmer_runtime::Module = compile_module(hash.clone(), code, config)?;
-        let artifact = compiled.cache().map_err(|_e| {
-            VMError::CacheError(format!("Cannot make cached artifact for {:?}", &code_hash))
-        })?;
+    match cache.get(key.as_slice()) {
+        Ok(serialized) => {
+            match serialized {
+                Some(serialized) => {
+                    // We got cached code or error from DB cache.
+                    let record = CacheRecord::try_from_slice(serialized.as_slice()).unwrap();
+                    let code = match record {
+                        CacheRecord::Error(err) => return Err(err),
+                        CacheRecord::Code(code) => code,
+                    };
+                    let artifact = Artifact::deserialize(code.as_slice())
+                        .map_err(|_e| VMError::CacheError(DeserializationError))?;
+                    unsafe {
+                        load_cache_with(artifact, compiler_for_backend(Backend::Singlepass).unwrap().as_ref())
+                            .map_err(|_e| VMError::CacheError(DeserializationError))
+                    }
+                }
+                None => {
+                    // Nothing found in cache, create new record.
+                    let compiled: wasmer_runtime::Module =
+                        compile_module(code, config).map_err(|e| {
+                            let record = CacheRecord::Error(e.clone());
+                            let e1 = cache.put(key.as_slice(), &record.try_to_vec().unwrap());
+                            if e1.is_err() {
+                                // That's fine, just cannot cache compilation error.
+                                println!("Cannot cache an error");
+                            }
+                            e
+                        })?;
+
+                    let artifact = compiled.cache().map_err(|_e| {
+                        let e = VMError::CacheError(SerializationError);
```

Comment:
Should we panic on this later? It seems like serialization errors shouldn't happen, so we shouldn't cache this error

Reply:
Not sure, but reworked this place.